src/main/java/jp/ac/utokyo/rcast/karkinos/filter/SoftClipExtention.java

```
@@ -0,0 +1,181 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package jp.ac.utokyo.rcast.karkinos.filter;
+
+import jp.ac.utokyo.rcast.karkinos.utils.TwoBitGenomeReader;
+
+import htsjdk.samtools.Cigar;
+import htsjdk.samtools.CigarElement;
+import htsjdk.samtools.CigarOperator;
+import htsjdk.samtools.SAMRecord;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class SoftClipExtention {
+
+    //add 2020/12/17 for FFPE anneling near repeat, extends soft clipping
+    public static void extendSoftclip(SAMRecord sam, TwoBitGenomeReader tgr) {
+        try {
+            if(sam.getCigar().getCigarElements().size()>=2) {
+                leftExtention(sam, tgr);
+                rightExtention(sam, tgr);
+            }
+
+        }catch(Exception ex){
+            //could not extend soft clips
+            ex.printStackTrace();
+        }
+    }
+
+    private static void leftExtention(SAMRecord sam, TwoBitGenomeReader tgr) {
+        List<CigarElement> list = sam.getCigar().getCigarElements();
+        CigarElement first = list.get(0);
+        CigarElement second = list.get(1);
+        int extraCheckLen = 15;
+
+        if(isClipped(first)&&isM(second)&&second.getLength()>extraCheckLen){
+
+            int alignmentStart = sam.getAlignmentStart();
+            int miscount = 0;
+            int lastidx = 0;
+            int m = 0;
+            int clipidx = 0;
+            for(int n=alignmentStart;n<=alignmentStart+extraCheckLen;n++){
+                //
+                char refNuc = tgr._getGenomeNuc(n,true);
+                int idx = sam.getReadPositionAtReferencePosition(n)-1;
+                if(idx>=0 && idx<sam.getReadLength()){
```

`getReadPositionAtReferencePosition` returns 1-based inclusive position into the unclipped sequence.
For alignments that have a cigar like `nSkM...` where n > 0, k >= 16, these conditions must always be true.
We should throw an exception when they fail since such alignments are invalid.

---

Oh,this is a mistake. Thankyou for pointing out.Then it should be like
int idx = sam.getReadPositionAtReferencePosition(n)-1+first.getReadLength();